Clarify system of equations between different sub-clocks. (#3284)
modelica/ModelicaSpecification: chapters/synchronous.tex

\end{lstlisting}
\end{example}

\begin{example}

> Forbidding systems of equations involving different sub-clocks mean that the following is forbidden:

Reviewer: Grammar:
```suggestion
Forbidding systems of equations involving different sub-clocks means that the following is forbidden:
```

\begin{lstlisting}[language=modelica]
  Real x=sample(time, Clock(1,100));
  Real y;
  Real z=subSample(y, 2)+x;
equation
  z=0;
\end{lstlisting}
Here \lstinline!x! and \lstinline!z! are part of one sub-clock partition and \lstinline!y! of another.

> The variables \lstinline!y! and \lstinline!z! are part of a system of equations spanning two sub-clocks, that is not legal.

Reviewer: I'd find this more readable (with _allowed_ being the natural opposite of _forbidden_):
```suggestion
The variables \lstinline!y! and \lstinline!z! are part of a system of equations spanning two sub-clocks, which is not allowed.
```

However, it is not required that the sub-clocks can necessarily be sorted as shown by following legal example:
\begin{lstlisting}[language=modelica]
  Real x=sample(time, Clock(1,100));
  Real y=superSample(x, 2);
  Real z=subSample(y, 2)+x;
\end{lstlisting}
Here \lstinline!x! and \lstinline!z! are part of one sub-clock partition, and \lstinline!y! of another.

> The equations can be sorted, but not the sub-clock partitions.
> Note that systems of equations within one sub-clock are also legal.

Reviewer: Re-write with focus on equation systems?
```suggestion
The equations form three equation systems with one equation in each (hence trivially satisfying the requirement that only variables from one sub-clock partition are being solved).
The equation systems need to be solved in a strict order, but the first and last equation system belong to one sub-clock, while the second equation system belongs to another sub-clock.
This illustrates that there is no guarantee that the sub-clock partitions can be ordered in agreement with the equation systems.
Note that equation systems with more than one equation are also allowed in sub-clock partitions.
```

\end{lstlisting}
\end{example}

\begin{example}
Forbidding systems of equations involving different sub-clocks means that the following is forbidden:
\begin{lstlisting}[language=modelica]
  Real x=sample(time, Clock(1,100));
  Real y;
  Real z=subSample(y, 2)+x;
equation

>   z=0;

Reviewer: What is the purpose of including this equation while not having any equation for `y`?  I would find the example more useful if it was complete enough to just try it in a tool.

Author: It is complete - and hopefully failing in tools.
I will explain it in more detail

\end{lstlisting}
\end{example}

\begin{example}
Forbidding systems of equations involving different sub-clocks means that the following is forbidden:
\begin{lstlisting}[language=modelica]
  Real a;
  //Real x=a+z;
  Real y=superSample(a+z, 2);
  Real z;
equation
  a+z = sample(time, Clock(1,100));
  0 = subSample(y, 2)+a;
\end{lstlisting}
Here \lstinline!a! and \lstinline!z! are part of one sub-clock, and \lstinline!y! of another, and the system of equations involve both of them.

> Replacing \lstinline!a! by \lstinline!x-z! (and simplifying the equations) solves this issue and gives the next example.

Reviewer: The next example?

Author: Will reformulate.

Author: Have reformulated, the point was that the two examples were related in that way.

Reviewer: Good.  Resolving.
